ffi/src/ir.rs

```diff
@@ -0,0 +1,249 @@
+use alloc::{boxed::Box, string::String, vec::Vec};
+use core::mem::ManuallyDrop;
+
+use crate::{repr_c::CType, slice::SliceRef, FfiReturn, ReprC, Result};
+
+/// Trait for `#[repr(transparent)]` structs to convert between [`Self`] and [`Self::Inner`]
+///
+/// # Safety
+/// `Self` and `Self::Inner` must have the same memory layout, the same set of valid values,
+/// and calling convention
+pub unsafe trait Transparent: Sized {
+    /// Non-ZST field of transparent struct
+    type Inner: Ir;
+
+    /// Convert transparent struct into its non-ZST field
+    // NOTE: Deriving `into_inner` is always valid because T::Inner is a superset of T
+    fn into_inner(outer: Self) -> Self::Inner;
+
+    /// Recover transparent struct from its non-ZST field
+    // FIXME: We should force the transparent type to have `from_inner` and derive implementation
+    // only if requested specifically by the user (via macro attribute). The reason for this is
+    // that even though two types are equal on the byte level, they may have different semantics
+    // that cannot be proven when deriving `from_inner` (e.g. `NonZeroU8` is a wrapper for `u8`)
+    // Deriving implementation of this method cannot be guaranteed to be correct
+    unsafe fn from_inner(inner: Self::Inner) -> Self;
+}
+
+// TODO: Add bound `Ir` or `FfiType` if possible
+pub unsafe trait Transmute {
+    type ReprC: ReprC;
+    unsafe fn is_valid(source: Self::ReprC) -> FfiReturn;
+}
+
+pub trait NonTransmute {}
+
+/// Facilitates conversion to/from internal representation. The given type is either forwarded
+/// as is or wrapped into [`Transmutable`] IR marker struct
+pub trait Ir: Sized {
+    /// Internal representation of the `Self`. Should be set either to `Self` or one of the marker
+    /// struct derivatives. Marker structs represent logically disjoint categories of types in IR
+    type Type;
+
+    /// Convert from [`Self`] into [`Self::Type`]
+    fn into_ir(self) -> Self::Type;
+
+    /// Convert from [`Self::Type`] into [`Self`]
+    fn from_ir(source: Self::Type) -> Self;
+}
+
+pub trait IrVec: Ir {
+    fn into_ir(vec: Vec<Self>) -> Vec<Self::Type>;
+    fn from_ir(source: Vec<Self::Type>) -> Vec<Self>;
+}
+pub trait IrSliceRef: Ir {
+    fn into_ir(slice_ref: &[Self]) -> &[Self::Type];
+    fn from_ir(source: &[Self::Type]) -> &[Self];
+}
+pub trait IrSliceMut: Ir {
+    fn into_ir(slice_mut: &mut [Self]) -> &mut [Self::Type];
+    fn from_ir(source: &mut [Self::Type]) -> &mut [Self];
+}
+
+#[repr(C)]
+pub union Transmutable<T: Transmute, U: ReprC> {
```

Comment:
Do we assume that `T` and `U` have the same memory layout?

Reply:
yes, this is an invariant enforced in the `Transmute` trait